### ISSUE TITLE
thunderbird support?

### ISSUE THREAD

**Issue by pcause** [gorhill/uBlock#117]
I notice that adblock and a few others run in thunderbird.  i was wondering if we can get ublock0 to run in thunderbird as well.  Looking at the MDN docs it would see that you can add another target application section for thunderbird and it _might_ work.

**Comment by gorhill** [comment]
It's simple: a developer has to come in and see what needs to be done to make it work.

**Comment by M-Reimer** [comment]
I was trying to get this to work, but I fail with the uBlock internal API in several situations.

Blocking already works and I could live with something like "partial support" where I can at least get the settings pages to work, but I don't know exactly where I have to edit. I've already managed to get a tab with the settings opened, but error console prints stuff like this:

Error: TypeError: vAPI.cloud is undefined
Source File: chrome://ublock0/content/js/messaging.js
Line: 693

Thunderbird is different in many cases.

For example there, in fact, is only one active browser at a time. So there is no real "tabbed browsing". Every single item of the GUI (folder list, message pane, ...) only exists once and is recycled when someone switches between the tabs. Long story short: There is not one browser per tab but there is only one or at maximum two browser objects. When someone switches to another tab, then this one browser has to navigate to the last page, that was open on this tab.

I don't know how to get this concept into uBlock as everything seems to expect real tabbed browsing...

**Comment by pcause** [comment]
Thanks so much for doing this.  I am downloading now.

**Comment by flyingzebras** [comment]
@M-Reimer  @vgturtle127 Basic Thunderbird 38.3.0 support have been added in uBlock Origin 1.2.2b2: https://github.com/gorhill/uBlock/releases/tag/1.2.2b2. We need a working version of uBlock Origin for Thunderbird 31.8.0 though. That will work on Icedove 31.8.0 too.

**Comment by flyingzebras** [comment]
@gorhill What do you think?

Do you work on Thunderbird support?

**Comment by gorhill** [comment]
> We need a working version of uBlock Origin for Thunderbird 31.8.0 though. That will work on Icedove 31.8.0 too.

Somebody using Thunderbird 31.8.0 will have to contribute and test whether uBO works just fine on it.

**Comment by flyingzebras** [comment]
@M-Reimer Can you do that?

**Comment by vgturtle127** [comment]
Why do Thinderbird 31.8 / Icedove 31.8 need to be supported? Don't they have newer versions that it will support already? Are these ESR-type releases? I don't want to get in to the habit of supporting every minor version and every fork, and every re-brand, etc.

**Comment by gorhill** [comment]
Another question I have for @flyingzebras is why ask _somebody else_ to do the testing? If you want support for Thunderbird 31.8.0, why not test yourself if it works fine?

**Comment by M-Reimer** [comment]
@flyingzebras: gorhill wrote "Somebody using Thunderbird 31.8.0" and so I'm out ;) But I think you are a good candidate to do the testing :P Just ask if you need any help.

There is a good chance that I will create more pull requests, targeting Thunderbird, in future. There are still some things missing like the toolbar button and integration into context menu. But I'll only test them with the Thunderbird version, I have in daily use.

**Comment by flyingzebras** [comment]
Icedove 31.8.0 (based on Thunderbird ESR 31.8.0) is the version used in the package icedove in Trisquel GNU/Linux. I get this message when I try to install uBlock Origin 1.3.1 in Icedove: "uBlock Origin could not be installed because it is not compatible with Icedove 31.8.0." @M-Reimer @gorhill @vgturtle127

**Comment by M-Reimer** [comment]
So the first thing, you have to do, is to mark your version as compatible.

If you didn't do so far, you have to clone the repository, first:

`$ git clone https://github.com/gorhill/uBlock.git`

Now edit the following line:
https://github.com/gorhill/uBlock/blob/master/platform/firefox/install.rdf#L68

After doing so, change back to the root directory of the checked out sourcecode and run:

`$ tools/make-firefox.sh all`

After doing so, you'll find your XPI file somewhere below "dist" in the source directory.

**Comment by gorhill** [comment]
Just found out AMO let me override the version compatibility information from the RDF file. So will set the compatibility to 31.8.0 for the latest dev build (1.3.2b5). It will be a matter of downloading and installing to see if it works.

**Comment by flyingzebras** [comment]
@gorhill Will your decision affect Thunderbird (and Icedove) versions too so I can try it in Icedove 31.8.0?

**Comment by gorhill** [comment]
> Will your decision

There is no decision, I set the minimum Thunderbird version to 31.0 so that you can test it. I await feedback on whether it works fine or not.

**Comment by flyingzebras** [comment]
@gorhill Thank you.

**Comment by flyingzebras** [comment]
@gorhill Please let me know when I can download https://github.com/gorhill/uBlock/releases/download/1.3.2b5/uBlock0.firefox.xpi with that update.

**Comment by gorhill** [comment]
I posted the link above: it's the latest dev build on AMO.

**Comment by flyingzebras** [comment]
It was not changed in https://addons.mozilla.org/firefox/downloads/file/362806/ublock_origin-1.3.2b5-sm+tb+fx+an.xpi

**Comment by FrankNitty** [comment]
I have it installed on 38.3 and can get to the settings through the addon panel but I can't seem to get it on the toolbar. It is nowhere to be found in the customize buttons to add it anywhere.
Would love to test it for Thunderbird but I can't find it.

**Comment by gorhill** [comment]
> `<minVersion>38.3.0</minVersion>` was not changed in

I didn't say it was, I said AMO allows to override on AMO.

What matters is: does it install on Thunderbird 31.8?

**Comment by flyingzebras** [comment]
@gorhill https://addons.mozilla.org/firefox/downloads/file/362806/ublock_origin-1.3.2b5-sm+tb+fx+an.xpi did not install on Icedove 31.8.0.

**Comment by gorhill** [comment]
I added an Icedove pack in latest release: <https://github.com/gorhill/uBlock/releases/tag/1.3.2b5>.

**Comment by flyingzebras** [comment]
@gorhill Thank you, it installed this time. But I cannot see it other than on Add-ons as FrankNitty pointed out above.

**Comment by gorhill** [comment]
It's partial support, as [pointed out everywhere](https://github.com/gorhill/uBlock/issues/117#issuecomment-146240233).

**Comment by flyingzebras** [comment]
@M-Reimer Please read my conversation with gorhill. Whats next?

**Comment by M-Reimer** [comment]
Go to Addons-Manager and choose "Preferences" for "uBlock Origin". Scroll to the end where you find "Show Dashboard" and "Show Logger".

Currently there is no toolbar button to add to somewhere. The preferences page in Addons manager is the only way to get access to the settings page.

**Comment by FrankNitty** [comment]
Thanks @M-Reimer 

Hopefully that will be added in the future

**Comment by flyingzebras** [comment]
@gorhill @M-Reimer Wow, it works! =)

**Comment by pcause** [comment]
just installed on the latest thunderbird 42 release candidate. i can see toolbar buttons for other addons when I go to customize.  is this something you have left to do or is there an issue with implementing?

**Comment by M-Reimer** [comment]
It is not easy to implement this as Thunderbird is completely different to Firefox in many cases.

My primary goal was to have uBlock Origin available so I can filter out ads in RSS feeds. For this the toolbar button is not that important.

By adding the toolbar button many things will be available and will cause new problems, like element picker mode.

I've already written here: https://github.com/gorhill/uBlock/issues/117#issuecomment-149916679 that I'll try to add more features, but this will definitively not be ready for the next release, so at least for some time the official way to get to the settings page is via Addons Manager.

**Comment by berrythesoftwarecodeprogrammar** [comment]
ublock0 is now the 3rd most used thunderbird addon according to https://addons.mozilla.org/en-US/thunderbird/extensions/?sort=users

ive been using it for a while now and its good but requires using the logger or manually editing rules to add `mailbox-scheme` rules when wanting to whitelist certain email content as i have it all blocked by default.

the only other issue is the icon is really broken 99% of the time. the red part indicates what my ublock icon looks like usually:
![ub0thunderbird](https://cloud.githubusercontent.com/assets/6548571/14416681/462eefbe-fff0-11e5-8dc1-da846a246966.png)

**Comment by vgturtle127** [comment]
@berrythesoftwarecodeprogrammar Thunderbird 38.7.2 is out, so try that and see if that helps. What operating system are you using?

**Comment by berrythesoftwarecodeprogrammar** [comment]
i just updated to 38.7.2 and theres no change and i didnt expect there to be. its been looking like that for the last several versions of thunderbird (ever since i started using ublock on thunderbird). im using windows 7 x64

**Comment by vgturtle127** [comment]
@berrythesoftwarecodeprogrammar Weird, I haven't had that problem on Windows 7, 8, or 10. On any version of Thunderbird or uBlock.

Maybe try a re-installation of Thunderbird? I don't know if you store email locally or not, but back it up if you do.

**Comment by berrythesoftwarecodeprogrammar** [comment]
@vgturtle127 my icon issue is now fixed after updating to 45.0 :). interesting version jump

**Comment by vgturtle127** [comment]
@berrythesoftwarecodeprogrammar Good! :smile:

**Comment by Alpengreis** [comment]
It seems it's not possible to export uBO backup-files in Thunderbird. Any ideas?

**Comment by davidhedlund** [comment]
@Alpengreis Can you please file a new issue for this?

**Comment by stanlane** [comment]
uBO 1.13.10 and up lacks Thunderbird support. Will you add it back @gorhill?

**Comment by gorhill** [comment]
Thunderbird does not support webext, you will have to stick to 1.13.8, or install manually by downloading `uBlock0.firefox.xpi` from the releases section.

**Comment by stanlane** [comment]
@gorhill Thanks.

**Comment by stanlane** [comment]
@gorhill Why is this issue open then?

**Comment by gorhill** [comment]
Somebody else did the work, I declined to do this myself. I don't know whether the work was considered complete, it's not for me to decide.

**Comment by B00ze64** [comment]
Good day.

I understand M-Reimer did the work for Thunderbird?

@M-Reimer I am using TB38 (I cannot upgrade until some Theme I use is also upgraded) with uB0 1.13.8. Today TB crashed, so I went MENU -> HELP -> Troubleshooting-Information, which opens a new tab (ABOUT:Support I guess). IMMEDIATE crash. I tried in SafeMode and it works, so I started disabling AddOns. Turns out the problem is with uB0. If uB0 is enabled, then it crashes as soon as I open Troubleshooting-Information. This is pretty annoying as I am trying to see what makes TB crash sometimes (I think it is SS2DevProps.dll which Asus sound drivers inject into all processes). I would like to be able to use ABOUT:Support in TB without having to remember to disable uB0.

Any chance you can look at this?

Thank you.
Best Regards,

**Comment by B00ze64** [comment]
@M-Reimer Oops sorry, here is the crash report:
https://crash-stats.mozilla.com/report/index/301ef049-6b4e-4736-b7a8-13f5d0170909

**Comment by M-Reimer** [comment]
I cannot read crash reports. I think they can only really be interpreted by Mozilla's own developers.

Can you try disabling those drivers?

The Thunderbird support never worked perfectly. For example there is https://github.com/gorhill/uBlock/issues/897 that I can't fix as I miss background knowledge and don't have time to find out how uB0 works here in backend and how to make it interact with Thunderbird in the right way.

And I'm thinking of even switching away from Thunderbird at all as it misses some standards like CardDAV that I need for mobile synchronization. Just didn't find a better Mail client, so far.

**Comment by salim-b** [comment]
> And I'm thinking of even switching away from Thunderbird at all as it misses some standards like CardDAV that I need for mobile synchronization. Just didn't find a better Mail client, so far.

Pretty off-topic, but I thought it might be worth sharing: There is this add-on called [SOGo Connector](https://sogo.nu/download.html#/frontends) which adds the missing CardDAV synchronization functionality. It works with the default Thunderbird address book. I've been using it for about 2 years now to sync with my [Posteo.de](https://posteo.de/hilfe/wie-richte-ich-den-adressbuchabgleich-mit-thunderbird-ein) account, works like a charm.

Besides there's another add-on called [CardBook](https://addons.mozilla.org/de/thunderbird/addon/cardbook/) which also supports CardDAV but adds a completely new/separarte address book (which probably excells the built-in one feature-wise) but I haven't tested it personally.

**Comment by B00ze64** [comment]
Good day.

I tried disabling the Asus crap that's injected into all processes (I renamed the DLLs and rebooted and then checked to make sure the DLLs were not in the process) but it still crashes. The bug report says file ...\build\mozilla\gfx\layers\client/ClientLayerManager.cpp, line 767) so it appears to crash while painting or moving layers around; not sure how uB0 creates a problem with painting layers. Yup, about-scheme is whitelisted. Could be a gfx driver problem. How's it working for you when you go Menu -> Help -> Troubleshooting; works fine?

I added uB0 to TB because I read RSS feeds, and /some/ feeds open a TAB in Firefox (to Twitter) when I read them; it is very annoying as FF comes to the foreground. I can switch to something else I guess (ABP?) but I kinda like having the same in TB as I have in FF...

Regards,

**Comment by salim-b** [comment]
> How's it working for you when you go Menu -> Help -> Troubleshooting; works fine?

Yeah, works fine for me. But I'm using Ubuntu Linux (16.04 LTS) with Thunderbird 52.2.1. Sorry, I don't have any more advice concerning your problem.

**Comment by B00ze64** [comment]
I'm trying-out the latest uB0 (1.14.14) in TB (38). Seems to work however, the list of filters is not being updated, it still shows deprecated filter lists. On Firefox I went into settings first, before updating from 1.13.8 to 1.14.14, and the list of filters updated itself while I was still on 1.13.8. Is this new and normal behavior, or an issue with my old TB38?

**Comment by gorhill** [comment]
Support for uBO/legacy for Thunderbird will not happen at this point.

**Comment by anewuser** [comment]
Current Thunderbird users might want to follow this: #3698.
